src/Codeception/Module/REST.php

```diff
@@ -632,25 +632,41 @@ protected function execute($method, $url, $parameters = [], $files = [])
         // allow full url to be requested
         if (!$url) {
             $url = $this->config['url'];
+        } elseif (!is_string($url)) {
+            throw new ModuleException(__CLASS__, 'URL must be string');
         } elseif (strpos($url, '://') === false && $this->config['url']) {
             $url = rtrim($this->config['url'], '/') . '/' . ltrim($url, '/');
         }
 
         $this->params = $parameters;
 
-        $parameters = $this->encodeApplicationJson($method, $parameters);
         $isQueryParamsAwareMethod = in_array($method, self::QUERY_PARAMS_AWARE_METHODS, true);
 
-        if (is_array($parameters) || $isQueryParamsAwareMethod) {
-            if (!empty($parameters) && $isQueryParamsAwareMethod) {
-                if (strpos($url, '?') !== false) {
-                    $url .= '&';
-                } else {
-                    $url .= '?';
+        if ($isQueryParamsAwareMethod) {
+            if (!is_array($parameters)) {
+                throw new ModuleException(__CLASS__, $method . ' parameters must be passed in array format');
+            }
+        } else {
+            $parameters = $this->encodeApplicationJson($method, $parameters);
+
+            if (!is_string($parameters) && !is_array($parameters)) {
+                if ($parameters instanceof \JsonSerializable) {
+                    throw new ModuleException(__CLASS__, $method . ' parameters is JsonSerializable object, but Content-Type header is not set to application/json');
                 }
-                $url .= http_build_query($parameters);
+                throw new ModuleException(__CLASS__, $method . ' parameters must be array, string or object implementing JsonSerializable interface');
             }
+        }
+
+        if (is_array($parameters) || $isQueryParamsAwareMethod) {
             if ($isQueryParamsAwareMethod) {
+                if (!empty($parameters)) {
+                    if (strpos($url, '?') !== false) {
```

Comment:
At this point I have forgotten that I am in a function called 'execute'. Given the increasing number of validations, I think it would be better if you could extract the validations to a separate private method responsible only for that, and leave here what is really needed to 'execute' assuming that the data is already valid.

Reply:
The problem is that this method is the only place where various `send` methods converge.
Also methods with request body accept different parameter types than method without request body.

I moved invalid parameter handling to encodeApplicationJson method, it is the only place where JsonSerializable interface is used now.

Reply:
I get it, that makes sense. Thank you.